src/terminal/input/terminalInput.cpp

```diff
@@ -515,169 +162,442 @@ TerminalInput::OutputType TerminalInput::HandleKey(const INPUT_RECORD& event)
-    if (WI_IsAnyFlagSet(keyEvent.dwControlKeyState, ALT_PRESSED) && !WI_IsAnyFlagSet(keyEvent.dwControlKeyState, CTRL_PRESSED) && keyEvent.uChar.UnicodeChar != 0)
+    // If we've got nothing usable, we'll just return an empty string. The event
+    // has technically still been handled, even if it's an unmapped key.
+    return MakeOutput({});
+}
+
+TerminalInput::OutputType TerminalInput::HandleFocus(const bool focused) const
+{
+    if (!_inputMode.test(Mode::FocusEvent))
     {
-        return _makeEscapedOutput(keyEvent.uChar.UnicodeChar);
+        return MakeUnhandled();
     }
 
-    // Pressing the control key causes all bits but the 5 least
-    // significant ones to be zeroed out (when using ASCII).
-    // This results in Ctrl+Space and Ctrl+@ being equal to a null byte.
-    // Normally the C0 control code set only defines Ctrl+@,
-    // but Ctrl+Space is also widely accepted by most terminals.
-    // -> Send a "null input sequence" in that case.
-    // We don't need to handle other kinds of Ctrl combinations,
-    // as we rely on the caller to pretranslate those to characters for us.
-    if (!WI_IsAnyFlagSet(keyEvent.dwControlKeyState, ALT_PRESSED) && WI_IsAnyFlagSet(keyEvent.dwControlKeyState, CTRL_PRESSED))
-    {
-        const auto ch = keyEvent.uChar.UnicodeChar;
-        const auto vkey = keyEvent.wVirtualKeyCode;
+    return MakeOutput(focused ? _focusInSequence : _focusOutSequence);
+}
 
-        // Currently, when we're called with Ctrl+@, ch will be 0, since Ctrl+@ equals a null byte.
-        // VkKeyScanW(0) in turn returns the vkey for the null character (ASCII @).
-        // -> Use the vkey to alternatively determine if Ctrl+@ is being pressed.
-        if (ch == UNICODE_SPACE || (ch == UNICODE_NULL && vkey == LOBYTE(OneCoreSafeVkKeyScanW(0))))
+void TerminalInput::_initKeyboardMap() noexcept
+try
+{
+    auto defineKeyWithUnusedModifiers = [this](const auto keyCode, const auto sequence) {
+        for (auto m = 0; m < 8; m++)
+            _keyMap[VTModifier(m) + keyCode] = sequence;
+    };
+    auto defineKeyWithAltModifier = [this](const auto keyCode, const auto sequence) {
+        _keyMap[keyCode] = sequence;
+        _keyMap[Alt + keyCode] = L"\x1B" + sequence;
+    };
+    auto defineKeypadKey = [this](const auto keyCode, const auto prefix, const auto finalChar) {
+        _keyMap[keyCode] = fmt::format(FMT_COMPILE(L"{}{}"), prefix, finalChar);
+        for (auto m = 1; m < 8; m++)
+            _keyMap[VTModifier(m) + keyCode] = fmt::format(FMT_COMPILE(L"{}1;{}{}"), CSI, m + 1, finalChar);
+    };
+    auto defineEditingKey = [this](const auto keyCode, const auto parm) {
+        _keyMap[keyCode] = fmt::format(FMT_COMPILE(L"{}{}~"), CSI, parm);
+        for (auto m = 1; m < 8; m++)
+            _keyMap[VTModifier(m) + keyCode] = fmt::format(FMT_COMPILE(L"{}{};{}~"), CSI, parm, m + 1);
+    };
+    auto defineNumericKey = [this](const auto keyCode, const auto f) {

```

Comment:
I believe it would be better to make all of these lambda parameters explicit. This avoids them being templates and ensures we only have 1 instance of each of these.